vm/systemSmartContracts/liquidStaking.go

```diff
@@ -0,0 +1,244 @@
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const tokenIDKey = "tokenID"
+const noncePrefix = "n"
+
+type liquidStaking struct {
+	eei                      vm.SystemEI
+	sigVerifier              vm.MessageSignVerifier
+	delegationMgrSCAddress   []byte
+	liquidStakingSCAddress   []byte
+	endOfEpochAddr           []byte
+	gasCost                  vm.GasCost
+	marshalizer              marshal.Marshalizer
+	hasher                   hashing.Hasher
+	mutExecution             sync.RWMutex
+	liquidStakingEnableEpoch uint32
+	flagLiquidStaking        atomic.Flag
+}
+
+// ArgsNewLiquidStaking defines the arguments to create the liquid staking smart contract
+type ArgsNewLiquidStaking struct {
+	EpochConfig            config.EpochConfig
+	Eei                    vm.SystemEI
+	DelegationMgrSCAddress []byte
+	LiquidStakingSCAddress []byte
+	EndOfEpochAddress      []byte
+	GasCost                vm.GasCost
+	Marshalizer            marshal.Marshalizer
+	Hasher                 hashing.Hasher
+	EpochNotifier          vm.EpochNotifier
+}
+
+// NewLiquidStakingSystemSC creates a new liquid staking system SC
+func NewLiquidStakingSystemSC(args ArgsNewLiquidStaking) (*liquidStaking, error) {
+	if check.IfNil(args.Eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+	if len(args.DelegationMgrSCAddress) < 1 {
+		return nil, fmt.Errorf("%w for delegation manager sc address", vm.ErrInvalidAddress)
+	}
+	if len(args.EndOfEpochAddress) < 1 {
+		return nil, fmt.Errorf("%w for end of epoch address", vm.ErrInvalidAddress)
+	}
+	if len(args.LiquidStakingSCAddress) < 1 {
+		return nil, fmt.Errorf("%w for liquid staking sc address", vm.ErrInvalidAddress)
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, vm.ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, vm.ErrNilHasher
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, vm.ErrNilEpochNotifier
+	}
+
+	l := &liquidStaking{
+		eei:                      args.Eei,
+		delegationMgrSCAddress:   args.DelegationMgrSCAddress,
+		endOfEpochAddr:           args.EndOfEpochAddress,
+		liquidStakingSCAddress:   args.LiquidStakingSCAddress,
+		gasCost:                  args.GasCost,
+		marshalizer:              args.Marshalizer,
+		hasher:                   args.Hasher,
+		liquidStakingEnableEpoch: args.EpochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
+	}
+	log.Debug("liquid staking: enable epoch", "epoch", l.liquidStakingEnableEpoch)
+
+	args.EpochNotifier.RegisterNotifyHandler(l)
+
+	return l, nil
+}
+
+// Execute calls one of the functions from the delegation contract and runs the code according to the input
+func (l *liquidStaking) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	l.mutExecution.RLock()
+	defer l.mutExecution.RUnlock()
+
+	err := CheckIfNil(args)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !l.flagLiquidStaking.IsSet() {
+		l.eei.AddReturnMessage("liquid staking contract is not enabled")
```

Comment:
I think this is not backwards compatible if we end up having a transaction towards the liquid staking contract before we have this code up and running.

Reply:
We do have this backward compatible - it is resolved directly in systemVM. line 108-110. It was done for delegation contract first. And contract have activation. They implement CanUseContract()

Reply:
ok